zig-mode.el

```diff
@@ -117,6 +128,66 @@ If given a SOURCE, execute the CMD on it."
   (interactive)
   (zig--run-cmd "run" (file-local-name (buffer-file-name)) "-O" zig-run-optimization-mode))
 
+;; zig flymake
+
+(defvar zig--flymake-proc nil
+  "Current zig flymake process.")
+(defun zig-flymake (report-fn &rest _args)
+  "Zig flymake command, called by flymake-start."
+  (unless (executable-find "zig")
+    (error "Cannot find suitable zig executable"))
+  ;; Kill the last zig--flymake-proc, as flymake can spawn them indefinitely.
+  (when (process-live-p zig--flymake-proc)
+    (kill-process zig--flymake-proc))
+
+  (save-restriction
+    (widen)
+    (setq
+     zig--flymake-proc
+     (make-process
+      :name "zig-flymake" :noquery t :connection-type 'pipe
+      :buffer (generate-new-buffer "*zig-flymake*")
+      :command zig-flymake-command
+      :sentinel
+      (lambda (proc _event)
+        (when (memq (process-status proc) '(exit signal))
+          (unwind-protect
+              (if (eq proc zig--flymake-proc)
+                  (with-current-buffer (process-buffer proc)
+                    (goto-char (point-min))
+                    (let ((diags '()))
+                      (while (search-forward-regexp
+                              "^\\(.*.zig\\):\\([0-9]+\\):\\([0-9]+\\): \\(.*\\)$" nil t)
+                        (let* ((msg (match-string 4))
+                               (file (match-string 1))
+                               (beg-line (string-to-number (match-string 2)))
+                               (beg-col (string-to-number (match-string 3)))
+                               ;; Zig does not warn, only errors or notes.
+                               (type (if (string-match "^error" msg)
+                                         :error
+                                       :note))
+                               (diag (flymake-make-diagnostic
+                                      file
+                                      (cons beg-line beg-col)
+                                      ;; Errors will have properly recalculated ends
+                                      ;; as when flymake-make-diagnostic gots file as
+                                      ;; locus, it automatically calls flymake-diag-region
+                                      ;; when trying to convert line and col into beg end.
+                                      nil
+                                      type
+                                      msg)))
+                          (setq diags (cons diag diags))))
+                      (funcall report-fn diags)))
+                (flymake-log :warning "Canceling obsolete check %s"
+                             proc))
+            (kill-buffer (process-buffer proc)))))))))
+
+(defun zig--setup-flymake-backend ()
+  (add-hook 'flymake-diagnostic-functions 'zig-flymake nil t))
+
+(if zig-flymake
```

Comment:
Doing this once at the top level is not going to work out well because changing `zig-flymake` after loading the file will make no difference. Better to add the `zig--setup-flymake-backend` to the default value of `zig-mode-hook` (which might mean adding an explicit `defcustom` for `zig-mode-hook`, or simply omit it and suggest that users call `add-hook` themselves.

If doing that, then you'd rename the function to `zig-setup-flymake-backend` because it's now effectively a public function.